### PR TITLE
Add a machine readable reason to all error paths

### LINKED ISSUE #9080: Add problem ID and exit code for "Ensure your system has enough CPUs. The minimum allowed is 2 CPUs"
**Steps to reproduce the issue:** 

1. Run `minikube start` on a default Google Cloud Compute VM with 1 cpu with `--output=json`
2. Get the error `
{"data":{"exitcode":"64","message":"Ensure your  system has enough CPUs. The minimum allowed is 2 CPUs.\n"},"datacontenttype":"application/json","id":"507528d8-0e85-47bb-a2d1-08b41d868683","source":"https://minikube.sigs.k8s.io/","specversion":"1.0","type":"io.k8s.sigs.minikube.error"}
`

`gcloud` needs a machine readable error code that matches 1-to-1 with the message for this so it can understand what the error was without parsing the message.

### PR DESCRIPTION
**Pull request by tstromberg**
This gargantuan PR updates all error paths to provide a machine readable reason. This PR also significantly expands the dynamic range of exit codes, providing a framework to use most of the ~125 exit codes available in UNIX.

This replaces the previous "problem" error matching package with a new general purpose "reason" package that can be used directly from anywhere. As such, this PR gets rid of many attempts to provide output that looked similar to the problem package. Like the old problem package, the reason ID is propagated into both the JSON output and display output.

Unlike the previous problem implementation, reasons no longer require an error regexp, and now supports custom exit codes and styles.

Fixes #9080 

### Old (exit code 251, UNDEFINED)

```
💣  docker daemon out of memory. No space left on device
```

In JSON form:

```
{"data":{"exitcode":"507","message":"docker daemon out of memory. No space left on device\n"},"datacontenttype":"application/json","id":"e2edb0f5-f8df-47d3-af2f-a15f22145223","source":"https://minikube.sigs.k8s.io/","specversion":"1.0","type":"io.k8s.sigs.minikube.error"}
```

### New (exit code 26, ExInsufficientStorage)

```
❌  Exiting due to RSRC_DOCKER_STORAGE: Docker is out of disk space! (/var is at 99% of capacity)
💡  Suggestion: 

    Try at least one of the following to free up space on the device:
    
    1. Run "docker system prune" to remove unused docker data
    2. Increase the amount of memory allocated to Docker for Desktop via
    Docker icon > Preferences > Resources > Disk Image Size
    3. Run "minikube ssh -- docker system prune" if using the docker container runtime
🍿  Related issue: https://github.com/kubernetes/minikube/issues/9024
```

In JSON form:

```
{"data":{"advice":"Try at least one of the following to free up space on the device:\n\t\n\t\t\t1. Run \"docker system prune\" to remove unused docker data\n\t\t\t2. Increase the amount of memory allocated to Docker for Desktop via \n\t\t\t\tDocker icon \u003e Preferences \u003e Resources \u003e Disk Image Size\n\t\t\t3. Run \"minikube ssh -- docker system prune\" if using the docker container runtime","exitcode":"26","issues":"https://github.com/kubernetes/minikube/issues/9024","message":"Exiting due to RSRC_DOCKER_STORAGE: Docker is out of disk space! (/var is at 100%% of capacity)","name":"RSRC_DOCKER_STORAGE","url":""},"datacontenttype":"application/json","id":"e6d5030a-b2d4-4e03-84d1-b2db7a8020f9","source":"https://minikube.sigs.k8s.io/","specversion":"1.0","type":"io.k8s.sigs.minikube.error"}
```

### Old (exit status 64, EX_USAGE)

```
💥  Your Docker Desktop has less than 2 CPUs. Increase CPUs for Docker Desktop. 
	
	Docker icon > Settings > Resources > CPUs
				
				
📘  https://docs.docker.com/config/containers/resource_constraints/
💣  Ensure your Docker for Desktop system has enough CPUs. The minimum allowed is 2 CPUs.
```

### New (exit status 29, ExInsufficientCores)
				
```
⛔  Exiting due to RSRC_DOCKER_CORES: Docker Desktop has less than 2 CPUs configured, but Kubernetes requires at least 2 to be available
💡  Suggestion: 

    1. Click on "Docker for Desktop" menu icon
    2. Click "Preferences"
    3. Click "Resources"
    4. Increase "CPUs" slider bar to 2 or higher
    5. Click "Apply & Restart"
📘  Documentation: https://docs.docker.com/docker-for-mac/#resources
```
				
📘  https://docs.docker.com/config/containers/resource_constraints/
💣  Ensure your Docker for Desktop system has enough CPUs. The minimum allowed is 2 CPUs.

### Old

```
❗  'docker' driver reported an issue: "docker version --format {{.Server.Os}}-{{.Server.Version}}" exit status 1: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
💡  Suggestion: Start the Docker service
📘  Documentation: https://minikube.sigs.k8s.io/docs/drivers/docker/

💣  Failed to validate 'docker' driver
```

### New (exit status 63: ExProviderNotRunning)

```

💣  Exiting due to PROVIDER_DOCKER_NOT_RUNNING: "docker version --format -" exit status 1: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
💡  Suggestion: Start the Docker service
📘  Documentation: https://minikube.sigs.k8s.io/docs/drivers/docker/

```

### Old  (exit 78, EX_CONFIG)

```
❌  [KUBECONFIG_DENIED] failed to start node startup failed: Failed to update kubeconfig file.: writing kubeconfig: Error writing file /Users/tstromberg/.kube/config: writefile failed for /Users/tstromberg/.kube/config: open /Users/tstromberg/.kube/config: permission denied
💡  Suggestion: Run: 'chmod 600 $HOME/.kube/config'
⁉️   Related issue: https://github.com/kubernetes/minikube/issues/5714
```

### New (exit 37, ExHostPermission)

```
🚫  Exiting due to HOST_KUBECONFIG_PERMISSION: Failed kubeconfig update: writing kubeconfig: Error writing file /Users/tstromberg/.kube/config: open /Users/tstromberg/.kube/config: permission denied
💡  Suggestion: Run: 'sudo chown $USER $HOME/.kube/config && chmod 600 $HOME/.kube/config'
🍿  Related issue: https://github.com/kubernetes/minikube/issues/5714
```

### Old

```
❗  Your system has 32768MB memory but Docker has only 2240MB. For a better performance increase to at least 3GB.

	Docker for Desktop  > Settings > Resources > Memory
```

### New

```
💨  Docker Desktop has access to only 2240MiB of the 32768MiB in available system memory. Consider increasing this for improved performance.
💡  Suggestion: 

    1. Click on "Docker for Desktop" menu icon
    2. Click "Preferences"
    3. Click "Resources"
    4. Increase "Memory" slider bar to 3 GB or higher
    5. Click "Apply & Restart"
📘  Documentation: https://docs.docker.com/docker-for-mac/#resources
```

### CODE REVIEW

File: cmd/minikube/cmd/mount.go [hunk 1]
```diff
@@ -100,25 +104,25 @@ var mountCmd = &cobra.Command{
...
 			ip, err = cluster.HostIP(co.CP.Host)
 			if err != nil {
-				exit.WithError("Error getting the host IP address to use from within the VM", err)
+				exit.Error(reason.IfHostIP, "Error getting the host IP address to use from within the VM", err)
 			}
 		} else {
 			ip = net.ParseIP(mountIP)
 			if ip == nil {
-				exit.WithCodeT(exit.Data, "error parsing the input ip address for mount")
+				exit.Message(reason.IfMountIP, "error parsing the input ip address for mount")
```
[the remark this answers is not in the document]
Comment (priyawadhwa) [reply cmd/minikube/cmd/mount.go:120]: Ah so I got to the file where these funcs are later in the review -- I think `exit.Error` checks if we have a solution for the error via the regex matching?

[the remark this answers is not in the document]
Comment (tstromberg) [reply cmd/minikube/cmd/mount.go:120]: Correct. exit.Message is when there is a precise message to deliver, exit.Error is when there is a fatal error. The latter does match against the database of known errors for an enhanced message. 

I've adjusted the docstring to mention this.

File: pkg/minikube/registry/registry.go [hunk 2]
```diff
@@ -74,6 +74,7 @@ type StatusChecker func() State
 type State struct {
 	Installed        bool
 	Healthy          bool
+	Running          bool // the provider does not appear to be running
```
[the remark this answers is not in the document]
Comment (tstromberg) [reply pkg/minikube/registry/registry.go:77]: Fixed.

**Pull request merged**